dartsim/src/SDFFeatures_TEST.cc

```diff
@@ -95,16 +109,196 @@ World LoadWorld(const std::string &_world)
+
+  resolvedAxis.SetXyzExpressedIn("");
+  return resolvedAxis;
+}
+
+/////////////////////////////////////////////////
+/// Downstream applications, like ign-gazebo, use this way of world construction
+WorldPtr LoadWorldPiecemeal(const std::string &_world)
+{
+  auto engine = LoadEngine();
+  EXPECT_NE(nullptr, engine);
+  if (nullptr == engine)
+    return nullptr;
+
+  sdf::Root root;
+  const sdf::Errors &errors = root.Load(_world);
+  EXPECT_EQ(0u, errors.size()) << errors;
+
+  EXPECT_EQ(1u, root.WorldCount());
+  const sdf::World *sdfWorld = root.WorldByIndex(0);
+  EXPECT_NE(nullptr, sdfWorld);
+
+  sdf::World newWorld;
+  newWorld.SetName(sdfWorld->Name());
+  newWorld.SetGravity(sdfWorld->Gravity());
+  auto world = engine->ConstructWorld(newWorld);
+  if (nullptr == world)
+    return nullptr;
+
+  std::unordered_map<const sdf::Model *, ModelPtr> modelMap;
+  std::unordered_map<const sdf::Link *, LinkPtr> linkMap;
+
+  auto createModel = [&](const sdf::Model *_model,
+                         const sdf::Model *_parentModel = nullptr) {
+    sdf::Model newSdfModel;
+    newSdfModel.SetName(_model->Name());
+    newSdfModel.SetRawPose(ResolveSdfPose(_model->SemanticPose()));
+    newSdfModel.SetStatic(_model->Static());
+    newSdfModel.SetSelfCollide(_model->SelfCollide());
+
+    ModelPtr newModel;
+    if (nullptr != _parentModel)
+    {
+      auto it = modelMap.find(_parentModel);
+      ASSERT_TRUE(it != modelMap.end());
+      newModel = it->second->ConstructNestedModel(newSdfModel);
+    }
+    else
+    {
+      newModel = world->ConstructModel(newSdfModel);
+    }
+
+    EXPECT_NE(nullptr, newModel);
+    if (nullptr != newModel)
+    {
+      modelMap[_model] = newModel;
+    }
+  };
+
+  for (uint64_t i = 0; i < sdfWorld->ModelCount(); ++i)
+  {
+    const auto *model = sdfWorld->ModelByIndex(i);
+    createModel(model);
+    for (uint64_t nestedInd = 0; nestedInd < model->ModelCount(); ++nestedInd)
+    {
+      createModel(model->ModelByIndex(nestedInd), model);
+    }
+  }
+
+  for (auto [sdfModel, physModel] : modelMap)
+  {
+    for (uint64_t li = 0; li < sdfModel->LinkCount(); ++li)
+    {
+      const auto *link = sdfModel->LinkByIndex(li);
+      sdf::Link newSdfLink;
+      newSdfLink.SetName(link->Name());
+      newSdfLink.SetRawPose(ResolveSdfPose(link->SemanticPose()));
+      newSdfLink.SetInertial(link->Inertial());
+
+      auto newLink = physModel->ConstructLink(newSdfLink);
+      EXPECT_NE(nullptr, newLink);
+      if (nullptr == newLink)
+        return nullptr;
+
+      linkMap[link] = newLink;
+    }
+  }
+
+  for (auto [sdfLink, physLink] : linkMap)
+  {
+    for (uint64_t ci = 0; ci < sdfLink->CollisionCount(); ++ci)
+    {
+      physLink->ConstructCollision(*sdfLink->CollisionByIndex(ci));
+    }
+  }
+
+  for (auto [sdfModel, physModel] : modelMap)
+  {
+    for (uint64_t ji = 0; ji < sdfModel->JointCount(); ++ji)
+    {
+      const auto *sdfJoint = sdfModel->JointByIndex(ji);
```

Comment:
```suggestion
      const auto sdfJoint = sdfModel->JointByIndex(ji);
```
I don't think we need a `*` in the return type since we are using `auto` and `JointByIndex` returns a `Joint *`.

Reply:
Can we also add a check to make sure `sdfJoint` isn't `nullptr` before de-referencing it below?

Reply:
Done in https://github.com/ignitionrobotics/ign-physics/pull/220/commits/53ce2b79c4059185820f7d5ded29221a9cabcb2b